### PR TITLE
ChainerCVRPNDecode

### PR DESCRIPTION
**Pull request by okdshin**

### CODE REVIEW

File: runtime/ops/something.cc [hunk 1]
```diff
@@ -13,44 +15,207 @@ namespace runtime {
...
+                        const double roi_l_h = ah * std::exp(std::min(loc_l_h, k_exp_clip));
+                        const double roi_l_w = aw * std::exp(std::min(loc_l_w, k_exp_clip));
 
-    WARN_ONCE("TODO: Implement ChainerCVRPNDecode");
+                        // yxhw -> tlbr (top left, bottom right)
+                        const double roi_l_tly = std::max(roi_l_y - roi_l_h / 2.0, 0.0);
+                        const double roi_l_tlx = std::max(roi_l_x - roi_l_w / 2.0, 0.0);
+                        const double roi_l_bry = std::min(roi_l_y + roi_l_h / 2.0, in_shape_h);
+                        const double roi_l_brx = std::min(roi_l_x + roi_l_w / 2.0, in_shape_w);
+                        auto data = std::shared_ptr<void>(new float[4]{static_cast<float>(roi_l_tly),
```
Comment (shinh) [inline runtime/ops/something.cc]: I'm not sure, but cannot we avoid the redundant heap allocation by argsorting confidences first?

File: runtime/ops/something.cc [hunk 2]
```diff
@@ -13,44 +15,207 @@ namespace runtime {
...
-    for (int bi = 0; bi < batch_size; ++bi) {
-        std::vector<chainerx::Array> roi;
-        std::vector<chainerx::Array> conf;
-        for (int l = 0; l < num_scales; ++l) {
+std::vector<size_t> NonMaximumSuppression(std::vector<chainerx::Array> const& roi_l_list, double nms_threash, size_t nms_limit) {
+    std::vector<double> bbox_area_list(roi_l_list.size());
+    std::transform(roi_l_list.begin(), roi_l_list.end(), bbox_area_list.begin(), [](const chainerx::Array& roi) {
+        return calc_area(extract_tlbr(roi));
+    });
+    std::vector<size_t> selec_indices;
```
[the remark this answers is not in the document]
Comment (okdshin) [reply runtime/ops/something.cc]: This `selec` comes from [original CahinerCV's code](https://github.com/chainer/chainercv/blob/master/chainercv/utils/bbox/non_maximum_suppression.py#L75).
It must mean "selected". I will modify it.

File: runtime/ops/something.cc [hunk 3]
```diff
@@ -13,44 +15,207 @@ namespace runtime {
...
+                    }  // end for iar
+                }  // end for v
+            }  // end for u
 
-    const int64_t num_rois = 7;
-    chainerx::Array fake_rois = chainerx::Zeros({num_rois, 4}, chainerx::Dtype::kFloat32, inputs[0].device());
-    chainerx::Array fake_roi_indices = chainerx::Zeros({num_rois}, chainerx::Dtype::kInt32, inputs[0].device());
+            // reduce size of `conf_l` to `nms_limit_pre`
+            std::vector<size_t> cut_indices(k_l);
+            std::iota(cut_indices.begin(), cut_indices.end(), 0);
```
[the remark this answers is not in the document]
Comment (okdshin) [reply runtime/ops/something.cc]: I thought `argsort` hides some rooms to optimize. However, actually, there is no such rooms. I will modify to use `argsort`.

File: runtime/ops/something.cc (hunk 1 above)
Comment (okdshin) [reply runtime/ops/something.cc]: Yes.

**Comment by shinh** [comment]
Failing to compile?

**Comment by okdshin** [comment]
Sorry, I will check.

**Comment by shinh** [comment]
Never mind. I enabled the test by myself. Thanks!

**Pull request merged**